// resolve/test/resolver/multirepo/packages/package-b/index.js

--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 387
Content-Type: application/octet-stream
X-File-MD5: 7dc9db9270547f5a146287d6a4bd6d78
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/multirepo/package.json

{
  "name": "ljharb-monorepo-symlink-test",
  "private": true,
  "version": "0.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "postinstall": "lerna bootstrap",
    "test": "node packages/package-a"
  },
  "author": "",
  "license": "MIT",
  "dependencies": {
    "jquery": "^3.3.1",
    "resolve": "../../../"
  },
  "devDependencies": {
    "lerna": "^3.4.3"
  }
}

--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 1
Content-Type: application/octet-stream
X-File-MD5: 68b329da9893e34099c7d8ad5cb9c940
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/cup.coffee



--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 24
Content-Type: application/octet-stream
X-File-MD5: 5e98b57495bcdf630a21ec378db7b0e9
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/precedence/bbb.js

module.exports = '>_<';

--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 0
Content-Type: application/octet-stream
X-File-MD5: d41d8cd98f00b204e9800998ecf8427e
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/mug.coffee


--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 0
Content-Type: application/octet-stream
X-File-MD5: d41d8cd98f00b204e9800998ecf8427e
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/mug.js


--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 20
Content-Type: application/octet-stream
X-File-MD5: 0c1d9e1731bb3d71b0b7a15695bfab14
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/foo.js

module.exports = 1;

--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 42
Content-Type: application/octet-stream
X-File-MD5: c957144f31f2bda33e82b3d31c6a4bda
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/test/resolver/false_main/package.json

{
	"name": "false_main",
	"main": false
}

--boundary_.oOo._KIEShwZW2ROn69QAthiwfovKmcICRdFW
Content-Length: 7025
Content-Type: application/octet-stream
X-File-MD5: 8c33718b32e0b3041bc85e46e47e39b0
X-File-Mtime: 1707173029
X-File-Path: /Codestack/Level 2/Assignments/news-homepage-main/node_modules/resolve/lib/sync.js

var isCore = require('is-core-module');
var fs = require('fs');
var path = require('path');
var getHomedir = require('./homedir');
var caller = require('./caller');
var nodeModulesPaths = require('./node-modules-paths');
var normalizeOptions = require('./normalize-options');

var realpathFS = process.platform !== 'win32' && fs.realpathSync && typeof fs.realpathSync.native === 'function' ? fs.realpathSync.native : fs.realpathSync;

var homedir = getHomedir();
var defaultPaths = function () {
    return [
        path.join(homedir, '.node_modules'),
        path.join(homedir, '.node_libraries')
    ];
};

var defaultIsFile = function isFile(file) {
    try {
        var stat = fs.statSync(file, { throwIfNoEntry: false });
    } catch (e) {
        if (e && (e.code === 'ENOENT' || e.code === 'ENOTDIR')) return false;
        throw e;
    }
    return !!stat && (stat.isFile() || stat.isFIFO());
};

var defaultIsDir = function isDirectory(dir) {
    try {
        var stat = fs.statSync(dir, { throwIfNoEntry: false });
    } catch (e) {
        if (e && (e.code === 'ENOENT' || e.code === 'ENOTDIR')) return false;
        throw e;
    }
    return !!stat && stat.isDirectory();
};

var defaultRealpathSync = function realpathSync(x) {
    try {
        return realpathFS(x);
    } catch (realpathErr) {
        if (realpathErr.code !== 'ENOENT') {
            throw realpathErr;
        }
    }
    return x;
};

var maybeRealpathSync = function maybeRealpathSync(realpathSync, x, opts) {
    if (opts && opts.preserveSymlinks === false) {
        return realpathSync(x);
    }
    return x;
};

var defaultReadPackageSync = function defaultReadPackageSync(readFileSync, pkgfile) {
    var body = readFileSync(pkgfile);
    try {
        var pkg = JSON.parse(body);
        return pkg;
    } catch (jsonErr) {}
};

var getPackageCandidates = function getPackageCandidates(x, start, opts) {
    var dirs = nodeModulesPaths(start, opts, x);
    for (var i = 0; i < dirs.length; i++) {
        dirs[i] = path.join(dirs[i], x);
    }
    return dirs;
};

module.exports = function resolveSync(x, options) {
    if (typeof x !== 'string') {
        throw new TypeError('Path must be a string.');
    }
    var opts = normalizeOptions(x, options);

    var isFile = opts.isFile || defaultIsFile;
    var readFileSync = opts.readFileSync || fs.readFileSync;
    var isDirectory = opts.isDirectory || defaultIsDir;
    var realpathSync = opts.realpathSync || defaultRealpathSync;
    var readPackageSync = opts.readPackageSync || defaultReadPackageSync;
    if (opts.readFileSync && opts.readPackageSync) {
        throw new TypeError('`readFileSync` and `readPackageSync` are mutually exclusive.');
    }
    var packageIterator = opts.packageIterator;

    var extensions = opts.extensions || ['.js'];
    var includeCoreModules = opts.includeCoreModules !== false;
    var basedir = opts.basedir || path.dirname(caller());
    var parent = opts.filename || basedir;

    opts.paths = opts.paths || defaultPaths();

    // ensure that `basedir` is an absolute path at this point, resolving against the process' current working directory
    var absoluteStart = maybeRealpathSync(realpathSync, path.resolve(basedir), opts);

    if ((/^(?:\.\.?(?:\/|$)|\/|([A-Za-z]:)?[/\\])/).test(x)) {
        var res = path.resolve(absoluteStart, x);
        if (x === '.' || x === '..' || x.slice(-1) === '/') res += '/';
        var m = loadAsFileSync(res) || loadAsDirectorySync(res);
        if (m) return maybeRealpathSync(realpathSync, m, opts);
    } else if (includeCoreModules && isCore(x)) {
        return x;
    } else {
        var n = loadNodeModulesSync(x, absoluteStart);
        if (n) return maybeRealpathSync(realpathSync, n, opts);
    }

    var err = new Error("Cannot find module '" + x + "' from '" + parent + "'");
    err.code = 'MODULE_NOT_FOUND';
    throw err;

    function loadAsFileSync(x) {
        var pkg = loadpkg(path.dirname(x));

        if (pkg && pkg.dir && pkg.pkg && opts.pathFilter) {
            var rfile = path.relative(pkg.dir, x);
            var r = opts.pathFilter(pkg.pkg, x, rfile);
            if (r) {
                x = path.resolve(pkg.dir, r); // eslint-disable-line no-param-reassign
            }
        }

        if (isFile(x)) {
            return x;
        }

        for (var i = 0; i < extensions.length; i++) {
            var file = x + extensions[i];
            if (isFile(file)) {
                return file;
            }
        }
    }

    function loadpkg(dir) {
        if (dir === '' || dir === '/') return;
        if (process.platform === 'win32' && (/^\w:[/\\]*$/).test(dir)) {
            return;
        }
        if ((/[/\\]node_modules[/\\]*$/).test(dir)) return;

        var pkgfile = path.join(maybeRealpathSync(realpathSync, dir, opts), 'package.json');

        if (!isFile(pkgfile)) {
            return loadpkg(path.dirname(dir));
        }

        var pkg = readPackageSync(readFileSync, pkgfile);

        if (pkg && opts.packageFilter) {
            // v2 will pass pkgfile
            pkg = opts.packageFilter(pkg, /*pkgfile,*/ dir); // eslint-disable-line spaced-comment
        }

        return { pkg: pkg, dir: dir };
    }

    function loadAsDirectorySync(x) {
        var pkgfile = path.join(maybeRealpathSync(realpathSync, x, opts), '/packa
